src/ca.c

```diff
@@ -2736,625 +2419,770 @@ static int ciplus_app_upgr_message(void *arg, uint8_t slot_id,
     return 0;
 } /* works now, be careful!!! just in case upgrade disabled */
 
-static int ciplus_app_ai_message(void *arg, uint8_t slot_id,
-                                 uint16_t session_number, uint32_t resource_id,
-                                 uint8_t *data, uint32_t data_length) {
-    ca_device_t *d = arg;
-    ca_slot_info_t info;
-    int fd = d->fd;
+static int ca_send_datetime(ca_device_t *d) {
+    unsigned char msg[6];
+    time_t tv = time(NULL);
+    uint16_t mjd = tv / 86400 + 40587; // mjd 01.01.1970 is 40587
+    tv %= 86400;
+    uint8_t hh = tv / 3600;
+    tv %= 3600;
+    uint8_t mm = tv / 60;
+    tv %= 60;
+    uint8_t ss = tv;
+
+    msg[0] = 5; // not using offset
+    msg[1] = (mjd >> 8) & 0xff;
+    msg[2] = mjd & 0xff;
+    msg[3] = ((hh / 10) << 4) | (hh % 10);
+    msg[4] = ((mm / 10) << 4) | (mm % 10);
+    msg[5] = ((ss / 10) << 4) | (ss % 10);
+    int session_number =
+        find_session_for_resource(EN50221_APP_DATETIME_RESOURCEID);
+    ca_write_apdu_session(d, session_number, TAG_DATE_TIME, msg, 6);
+    d->datetime_next_send = tv + d->datetime_response_interval;
+    LOG("Sending time to CA %d, response interval: %d", d->id,
+        d->datetime_response_interval)
+    return 0;
+}
 
-    if (data_length < 3) {
-        LOG("Received short data");
-        return -1;
-    }
-    uint32_t tag = (data[0] << 16) | (data[1] << 8) | data[2];
+int APP_empty(struct ca_device *d, int resource, uint8_t *buffer, int len) {
+    LOG("Got unhandled resource %06X", resource);
+    return 0;
+}
 
-    LOG("RECV ciplus AI msg slot_id %u, session_num %u, resource_id %x tag %x",
-        slot_id, session_number, resource_id, tag);
+static int APP_RM_create(ca_device_t *d, int session_number, int resource_id) {
+    LOG("--------------------S_SCALLBACK_REASON_CAMCONNECTED---------"
+        "APP_RM_RESOURCEID-------------------------");
+    ca_write_apdu(d, TAG_PROFILE_ENQUIRY, NULL, 0);
+    return 0;
+}
 
-    switch (tag) {
-    case CIPLUS_TAG_APP_INFO:
-        //		hexdump(data,data_length);
-        return en50221_app_ai_message(d->ai_resource, slot_id, session_number,
-                                      resource_id, data, data_length);
+int APP_RM_handler(struct ca_device *d, int resource, uint8_t *buffer,
+                   int len) {
+    int resource_ids[100], rlen;
+    switch (resource) {
+    case TAG_PROFILE:
+        ca_write_apdu(d, TAG_PROFILE_CHANGE, NULL, 0);
         break;
-    case CIPLUS_TAG_CICAM_RESET:
-        //		hexdump(data,data_length);
-        if (ioctl(fd, CA_RESET, &info))
-            LOG_AND_RETURN(0, "%s: Could not reset ca %d", __FUNCTION__, d->id);
-        return 1;
+    case TAG_PROFILE_ENQUIRY:
+        // do we really need this to force CI only mode ?
```

Comment:
I'm not sure but I don't think we have to - CI CAMs should simply ignore whatever unknown resources the host declares

Reply:
This is mostly for forcing using CI (not CI+) from the host side (if for example the certs are missing)

Reply:
True